src/utils/stream.ts

```diff
@@ -0,0 +1,399 @@
+}
+
+function isStreamSyncHistoryEnd(data: IStreamSyncHistoryData): data is IStreamSyncHistoryEnd {
+  return data.type === 'stream:history:end';
+}
+
+function isStreamSyncHistoryError(data: IStreamSyncHistoryData): data is IStreamSyncHistoryError {
+  return data.type === 'stream:history:error';
+}
+
+/**
+ * Load addresses in a CPU intensive way
+ * This only contemplates P2PKH addresses for now.
+ */
+export function loadAddressesCPUIntensive(
+  startIndex: number,
+  count: number,
+  xpubkey: string,
+  networkName: string
+): [number, string][] {
+  const addresses: [number, string][] = [];
+  const stopIndex = startIndex + count;
+  const network = new Network(networkName);
+  const hdpubkey = new HDPublicKey(xpubkey);
+
+  for (let i = startIndex; i < stopIndex; i++) {
+    const key = hdpubkey.deriveChild(i);
+    addresses.push([i, new BitcoreAddress(key.publicKey, network.bitcoreNetwork).toString()]);
+  }
+
+  return addresses;
+}
+
+export function generateStreamId() {
+  return Math.random().toString(36).substring(2, 15);
+}
+
+export async function xpubStreamSyncHistory(
+  startIndex: number,
+  _count: number,
+  storage: IStorage,
+  connection: FullnodeConnection,
+  shouldProcessHistory: boolean = false
+) {
+  let firstIndex = startIndex;
+  const scanPolicyData = await storage.getScanningPolicyData();
+  if (isGapLimitScanPolicy(scanPolicyData)) {
+    if (startIndex !== 0) {
+      const { lastLoadedAddressIndex } = await storage.getWalletData();
+      firstIndex = lastLoadedAddressIndex + 1;
+    }
+  }
+  await streamSyncHistory(
+    firstIndex,
+    storage,
+    connection,
+    shouldProcessHistory,
+    HistorySyncMode.STREAM_XPUB
+  );
+}
+
+export async function manualStreamSyncHistory(
+  startIndex: number,
+  _count: number,
+  storage: IStorage,
+  connection: FullnodeConnection,
+  shouldProcessHistory: boolean = false
+) {
+  let firstIndex = startIndex;
+  const scanPolicyData = await storage.getScanningPolicyData();
+  if (isGapLimitScanPolicy(scanPolicyData)) {
+    if (startIndex !== 0) {
+      const { lastLoadedAddressIndex } = await storage.getWalletData();
+      firstIndex = lastLoadedAddressIndex + 1;
+    }
+  }
+  await streamSyncHistory(
+    firstIndex,
+    storage,
+    connection,
+    shouldProcessHistory,
+    HistorySyncMode.STREAM_MANUAL
+  );
+}
+
+/**
+ * Start a stream to sync the history of the wallet on `storage`.
+ * Since there is a lot of overlap between xpub and manual modes this method was created to accomodate both.
+ * @param {number} startIndex Index to start loading addresses
+ * @param {IStorage} storage The storage to load the addresses
+ * @param {FullnodeConnection} connection Connection to the full node
+ * @param {boolean} shouldProcessHistory If we should process the history after loading it.
+ * @param {HistorySyncMode} mode The mode of the stream
+ * @returns {Promise<void>}
+ */
+export async function streamSyncHistory(
```

Comment:
I think this method is "doing too much" which makes it too complex and error prone.

Reply:
Yes, but I kept the funtion implementation to be consistent with our existing `syncHistory`.
I think once we have a functional implementation and integration testing we could add a refactor task to make these class based implementations like you suggested.

What do you think?

Reply:
I'm ok moving forward like this if we need to speed up the release of this feature. Otherwise, I guess we can create a new structure that is transparent to the caller regardless of the sync mode. I don't necessarily think we need a class based implementation. My feeling is that the current implementation is confusing and error prone.

Reply:
Made some changes to move the responsabilities of the client to a stream manager instance and the `streamSyncHistory` now only makes use of the manager to start, receive and cleanup streamming resources.

Reply:
I guess it's way better. Good job! Thanks!